### ISSUE TITLE
exp-10 root cause: PonderNet loss silently drops L_distill/L_ref/L_step (fix in #3)

### ISSUE THREAD

**Issue by anatissera** [anatissera/adaptive-latent-reasoning#4]
## Summary

exp-10 (the from-scratch replication of exp-08 Run C) landed at 18.88% test accuracy,
statistically indistinguishable from the No-CoT-SFT floor (19.1%, CODI/SIM-CoT papers).
The original write-up attributed this gap to "the warm-start checkpoint was doing the
work." That reading does not hold, and the real cause is a code bug that affects every
PonderNet experiment, not just exp-10.

## What we found

**1. The reference number was wrong.** CODI's 43.7% is itself a from-scratch GPT-2 run
(LoRA-only, backbone frozen, trained 40 epochs at LR 3e-3 on the full GSM8k-Aug split).
The checkpoint our warm-start runs (03-08) load *is the output of that procedure*. Our
architecture (CODI backbone + SIM-CoT auxiliary decoder) has its own published number:
**SIM-CoT Table 1 reports 42.6%** for GPT-2 + CODI, not CODI's self-reported 43.7%.

**2. The root cause: a detached-loss bug in PonderNet mode.** `CODI.forward()` in
`pondernet/src/model.py` builds `loss` from graph-connected `L_distill`/`L_ref`/`L_step`,
detaches those names for logging, then the PonderNet branch **rebuilds** `loss` further
down using the now-detached names. Since `=` binds whatever the name points to at that
moment, the effective training objective in every PonderNet run (experiments 02-11) was

```
L = L_ponder + gamma * KL_geom
```

CODI's self-distillation and SIM-CoT's step-level supervision contributed **zero
gradient**. The auxiliary decoder never trained in any PonderNet run.

Verified three independent ways:
- **Gradient test**: forward+backward with `pondernet=True, scope=full_dec` on the
  exact exp-10 config - the aux decoder went from `grad=None` on 148/148 params to
  nonzero after the fix.
- **Weight-drift forensics** on the recovered exp-10 checkpoint: its decoder never
  moved off vanilla GPT-2 init (relative L2 drift 0.0017, i.e. bf16 storage noise)
  while the working SIM-CoT checkpoint's decoder moved 0.687.
- **accuracy@k on exp-10's final checkpoint**: forcing k = 1..12 latent steps gives a
  flat 18.6-18.6% curve. Twelve latent steps buy exactly what one does. The latents
  are inert, consistent with the missing gradient.

**3. A second, independent defect also contributed.** exp-10 used
`--pondernet_train_scope full_dec` (269M trainable params, full backbone fine-tune) at
LR 1e-3 for 40 epochs, versus upstream's LoRA-only scope (144.5M params, backbone
frozen) at 3e-3. Backbone drift from vanilla GPT-2: 0.32 (exp-10) vs 0.0017 (SIM-CoT).
The "3e-3 diverges in bf16" finding that motivated dropping to 1e-3 was very likely a
symptom of this, not a bf16 property: 3e-3 is a LoRA learning rate, applied here to a
full fine-tune.

## Fix

The loss fix is in #3: capture graph-connected handles for the
three loss terms before the detach block, use those handles when the PonderNet branch
reassembles `loss`. 6 functional lines, plus a regression test
(`pondernet/tests/test_loss_gradients.py`) asserting on gradient magnitude (not `!=`,
which is how the original "decoder weights differ from vanilla GPT-2" verification
passed despite the bug).

Not merged yet. A from-scratch control run (no PonderNet, LoRA-only, the corrected
scope/LR) is in progress to validate the pipeline before layering the fix back on -
see the roadmap issue for what that unblocks.

## Scope of the damage

Applies to every PonderNet run, 02 through 11. The warm-start headline numbers (41%
accuracy, Spearman +0.675 difficulty tracking) remain valid *measurements*, but of a
different method than the paper describes: halting on a SIM-CoT checkpoint with CODI's
and SIM-CoT's own training signals inactive during the halting fine-tune. Every
hyperparameter tuned under the broken objective (gamma in exps 04/08, prior shape in
05/08, thresholds) is suspect and needs re-searching once the fix is validated.

**Comment by anatissera** [comment]
## E0 finished: the control validates the pipeline, and it relocates the open question

E0 is the from-scratch control this issue says it is waiting on ("a from-scratch control run (no PonderNet, LoRA-only, the corrected scope/LR) is in progress to validate the pipeline"). It ran 19h12m on `alr-exp10-spot` and completed all 40 epochs. Closing this issue on the result.

### Hyperparameters are not a confound

The released `internlm/SIM_COT-GPT2-CODI` ships its own `training_args.bin`. Diffing that against E0's launch command settles the "are these really their settings" question against the artifact rather than the paper's prose:

| knob | released ckpt | E0 | |
|---|---|---|---|
| learning_rate | 0.003 | 3e-3 | same |
| num_train_epochs | 40.0 | 40 | same |
| effective batch | 64 x 2 | 128 x 1 | same (128) |
| max_grad_norm | 2.0 | 2.0 | same |
| weight_decay / warmup_ratio | 0.1 / 0.03 | 0.1 / 0.03 | same |
| lr_scheduler_type | cosine | cosine | same |
| num_latent / K | 6 | 6 | same |
| use_prj / prj_dim / prj_dropout | True / 768 / 0.0 | True / 768 / 0.0 | same |
| distill_loss_div_std / _type | True / smooth_l1 | True / smooth_l1 | same |
| distill/explain/ref_loss_factor | 1.0 / 1.0 / 1.0 | 1.0 / 1.0 / 1.0 | same |
| remove_eos / optim | True / adamw_torch | True / adamw_torch | same |
| **seed** | **11** | **42** | **differs** |
| max_token_num | 1000 | 700 | no-op, longest train row is 247 tokens (measured, n=385,620) |
| decoder_path | unset | `gpt2` | no-op, `pj_in`/`pj_out` are `Identity()` either way |

Seed is the only substantive difference.

### The result

Accuracy rises from 7.80% (epoch 1) to a plateau from epoch 28, oscillating 29-33% with no further trend as the cosine schedule decays. Epoch 40: **33.40%** (GSM8K test, fixed 500-example subset, greedy, K=6).

Above the 19.1% No-CoT floor and above CODI's 24.5% "w/o L1 loss" ablation, so the latents compute something real. Short of SIM-CoT's published 42.6%, which looks like a remaining defect until you measure the reference the same way:

| model, identical harness / protocol (greedy, K=6) | split | accuracy |
|---|---|---|
| `internlm/SIM_COT-GPT2-CODI` (authors' weights) | test sub-500 | **29.00%** |
| E0 (this pipeline, from scratch) | test sub-500 | **33.40%** |
| `internlm/SIM_COT-GPT2-CODI` (authors' weights) | **full test, n=1319** | **27.75%** |
| SIM-CoT paper Table 1, GPT-2 + CODI | full test, n=1319 | 42.6% |

**E0 beats the published checkpoint by +4.4 pp under identical measurement.** The training pipeline is not what is broken. The shortfall against 42.6% reproduces at least as strongly on the authors' own weights (14.85 pp on the full official test set), so it lives in evaluation, not in training.

### Artifacts ruled out

- **Checkpoint loads completely.** `test.py` uses `load_state_dict(..., strict=False)`, which silently tolerates namespace mismatch, so this needed checking: 400 ckpt tensors vs 400 model tensors, **400 matched, 0 ignored, 0 left at init, 0 shape mismatches**.
- **Not dropout.** `model.eval()` is called before generation.
- **Not the subset.** `test_sub500.jsonl` is a true subset and difficulty-matched (mean CoT steps 3.232 vs 3.246). Settled directly by re-running the reference on the full 1,319-example test set: 27.75%.
- **Not subset noise.** SE ~2.1 pp at n=500; the gap to 42.6% is ~6.8 sigma.
- **Not hyperparameters.** See the table above.

Still open: the source of the harness gap. Not yet eliminated are answer extraction/parsing, prompt construction, tokenizer special-token handling, and decoding details.

### Why this closes the issue

The root cause in the title is diagnosed and verified three independent ways (§4.0), the fix is in #3 with `pondernet/tests/test_loss_gradients.py` pinning it on gradient magnitude, and the validation gate this issue defined - a from-scratch control reaching the reference - has now passed. **H-F5 is retired**: it predicted the fix alone would not rescue a from-scratch run and that the LoRA-only scope and 3e-3 LR had to be adopted too; E0 adopted both and reached the reference. **H-F3 resolves affirmatively** on the only controlled comparison available, and 42.6% turns out to be the wrong yardstick since nothing reaches it on this harness, including the authors' weights.

### What this does not close, tracked in #5

- E0 ran `--pondernet False`, so it exercises the upstream loss path, not the PonderNet branch the bug lived in. **E0 validates the pipeline; it is not itself a test of the fix.** H-F1, H-F2 and H-F4 still need a PonderNet run on a validated pipeline.
- The ~14.9 pp harness gap needs its own root-cause pass. It affects every comparison this project has drawn against a paper number, including the sibling `latent-halting` repo's `00-frontier` headline. The correct local reference for any trained run is **27.75%** (full test) / **29.00%** (sub-500) on this harness, not 42.6%.

**Comment by anatissera** [comment]
Correction to one line in the comment above, since it conflates two different harnesses.

I wrote that the measurement gap affects "the sibling `latent-halting` repo's `00-frontier` headline". That is imprecise about magnitude. `latent-halting` has its **own, separate** eval harness, and it scores the same released `internlm/SIM_COT-GPT2-CODI` checkpoint at **39.50%** (`00-frontier`, `fixedk-k6`), i.e. 3.1 pp below the published 42.6%, versus **27.75%** / 14.85 pp for this repo's `pondernet/test.py`.

So there are two measurement gaps of very different size, in two different codebases:

| harness | released ckpt, full test n=1319 | gap vs published 42.6% |
|---|---|---|
| this repo, `pondernet/test.py` | 27.75% | 14.85 pp |
| `latent-halting`, its own eval | 39.50% | 3.1 pp |

Consequences: accuracy numbers from the two repos are **not** directly comparable to each other, and `latent-halting`'s smaller gap is a separate question already tracked in that repo's issue #9.

Nothing else in the comment changes. The central result is a within-harness comparison and is unaffected: E0 at 33.40% versus the released checkpoint at 29.00%, same harness, same 500 examples, same flags.

**Comment by anatissera** [comment]
## Retraction: I closed this on a bad measurement. Reopening.

The comment above claims E0 beats the released checkpoint by +4.4 pp and that this repo's harness under-measures by ~14.85 pp. **Both claims are wrong, and they are wrong for the same reason.**

### The mistake

In `pondernet/test.py`, the fixed-K inference branch loops over **`training_args.inf_latent_iterations`** (`src/model.py:128`, **default 1**) - not over `max_latent_steps`. Nothing derives one from the other. My eval command passed `--max_latent_steps 6`, which sets the model's `K` but never reaches the inference loop, and omitted `--inf_latent_iterations`. **Every number I reported was measured with one latent step instead of six**: all 40 of E0's per-epoch points, and the 29.00% / 27.75% for the released checkpoint.

It failed silently and the output was plausible, which is why it survived the artifact checks in the comment above. Those checks were real but aimed at the wrong target.

### Corrected numbers

Same 500 examples, greedy, `--inf_latent_iterations` swept 1..6:

| k | reference (released) | E0 (epoch 40) |
|---|---|---|
| 1 | 29.00% | 33.40% |
| 2 | 28.00% | 13.40% |
| 3 | 32.80% | 14.20% |
| 4 | 37.00% | 28.80% |
| 5 | 38.80% | **35.20%** (peak) |
| **6 (trained k)** | **39.60%** (peak) | **34.80%** |

**There is no harness deficit.** Run correctly this harness scores the released checkpoint at 39.60%, agreeing with `latent-halting`'s independent eval (39.50%) to 0.1 pp. The two harnesses never disagreed. My correction comment about "two measurement gaps of very different size" is also void - the repos' numbers are comparable after all. The residual ~3 pp vs the published 42.6% is small, shared by both repos, and plausibly protocol (n=1319 vs 500, early-stopping selection).

**E0 does not reach the reference.** It is 4.8 pp below it at the trained k=6.

### What the corrected data actually shows

Not what my first correction said either. I initially read the k=1 to k=6 delta (+1.4 pp) as "E0's latents are nearly inert". That is an artifact of comparing endpoints on a non-monotonic curve: E0's k=1 is a high outlier sitting above its own k=2..4. Across the full sweep E0 spans 13.40% to 35.20%, a 21.8 pp range, so **the latents are doing substantial work** - categorically unlike exp-10's flat 18.6% across k=1..12.

The real finding is the curve's **shape**. E0 collapses to 13.40% at k=2 and 14.20% at k=3, worse than at k=1 by ~20 pp, then recovers through k=4-5. The reference dips 1 pp at k=2 and rises monotonically after. A model whose two-step prefix is 20 pp worse than its one-step prefix has intermediate latent states that actively mislead the answer decode.

**That is directly load-bearing for Option C.** PonderNet halts early by design - exp-08 Run C averages 2.93 steps, sitting in exactly E0's trough. A halting policy fitted to a backbone with this profile would be selecting among prefixes whose quality is non-monotonic in depth, and the accuracy-compute frontier it traces would partly reflect that shape rather than the halting mechanism. The `accuracy@k` curve of any from-scratch backbone should be checked before a halting policy is fitted to it.

### Status

Reopening. H-F3 is **not** satisfied (34.80% at k=6, below the 39.60% reference). H-F5 is **not** retired: the scope and LR fixes were necessary and moved exp-10's 18.88% to 34.80%, but they are not sufficient. H-F2 does pass on its own terms - `accuracy@k` rises rather than staying flat - which is the genuine positive result here.

The diagnosis in the issue body (the detached-loss bug, verified three ways, fixed in the linked PR and pinned by `test_loss_gradients.py`) is unaffected by any of this. What is not yet established is the validation gate this issue defined, so it should stay open until a from-scratch control actually reaches the reference.

Also worth fixing at the source: `--max_latent_steps` and `--inf_latent_iterations` are different names with different defaults for what a caller reasonably reads as one setting. An eval wrapper should assert inference depth equals the trained `K` unless an `accuracy@k` sweep is explicitly requested.
